File: Week_03/G20200343040330/144.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
  if(!root) return [];
  const result = [];
  const _preorderTraversal = (node) => {
      if(!node) return
      result.push(node.val);
      if(node.left) {
          _preorderTraversal(node.left)
      }
      if(node.right) {
          _preorderTraversal(node.right)
      }
  }
  _preorderTraversal(root)
  return result;
};

File: Week_03/G20200343040338/144.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        def traversal(root: TreeNode, vals: List[int]) -> List[int]:
            if not root:
                return

            vals.append(root.val)
            traversal(root.left, vals)
            traversal(root.right, vals)

        vals = []
        traversal(root, vals)

        return vals
